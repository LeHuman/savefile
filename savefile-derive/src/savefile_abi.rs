use proc_macro2::{Ident, Literal, Span, TokenStream};
use quote::ToTokens;
use syn::{GenericArgument, ParenthesizedGenericArguments, PathArguments, ReturnType, Type, TypeParamBound};

fn emit_closure_helpers(
    version: u32,
    temp_trait_name: Ident,
    args: &ParenthesizedGenericArguments,
    ismut: bool,
    extra_definitions: &mut Vec<TokenStream>,
    fnkind: Ident,
) {
    let temp_trait_name_wrapper = Ident::new(&format!("{}_wrapper", temp_trait_name), Span::call_site());

    let mut formal_parameter_declarations = vec![];
    let mut parameter_types = vec![];
    let mut arg_names = vec![];

    for (arg_index, arg) in args.inputs.iter().enumerate() {
        let arg_name = Ident::new(&format!("x{}", arg_index), Span::call_site());
        formal_parameter_declarations.push(quote! {#arg_name : #arg});
        parameter_types.push(arg.to_token_stream());
        arg_names.push(arg_name.to_token_stream());
    }

    let ret_type;
    let ret_type_decl;

    if let ReturnType::Type(_, rettype) = &args.output {
        let typ = rettype.to_token_stream();
        ret_type = quote! {#typ};
        ret_type_decl = quote! { -> #typ };
    } else {
        ret_type = quote! { () };
        ret_type_decl = quote! {};
    }

    let version = Literal::u32_unsuffixed(version);

    let mutsymbol;
    let mutorconst;
    if ismut {
        mutsymbol = quote! {mut};
        mutorconst = quote! {mut};
    } else {
        mutsymbol = quote! {};
        mutorconst = quote! {const};
    }

    let expanded = quote! {

        #[savefile_abi_exportable(version=#version)]
        pub trait #temp_trait_name {
            fn docall(& #mutsymbol self, #(#formal_parameter_declarations,)*) -> #ret_type;
        }

        struct #temp_trait_name_wrapper<'a> {
            func: *#mutorconst (dyn for<'x> #fnkind( #(#parameter_types,)* ) #ret_type_decl +'a)
        }
        impl<'a> #temp_trait_name for #temp_trait_name_wrapper<'a> {
            fn docall(&#mutsymbol self, #(#formal_parameter_declarations,)*) -> #ret_type {
                unsafe { (&#mutsymbol *self.func)( #(#arg_names,)* )}
            }
        }

    };
    extra_definitions.push(expanded);
}

enum ArgType {
    PlainData(TokenStream),
    Reference(TokenStream),
    SliceReference(TokenStream),
    TraitReference(Ident, bool /*ismut*/),
    BoxedTrait(Ident),
    Fn(
        Ident,       /*Name of temporary trait generated to be able to handle Fn* as dyn TemporaryTrait. */
        TokenStream, /*full closure definition (e.g "Fn(u32)->u16")*/
        Vec<Type>,   /*arg types*/
        bool,        /*ismut*/
    ),
}

pub(crate) struct MethodDefinitionComponents {
    pub(crate) method_metadata: TokenStream,
    pub(crate) callee_method_trampoline: TokenStream,
    pub(crate) caller_method_trampoline: TokenStream,
}

fn parse_type(
    version: u32,
    arg_name: &Ident,
    typ: &Type,
    method_name: &Ident,
    name_generator: &mut impl FnMut() -> String,
    extra_definitions: &mut Vec<TokenStream>,
    is_reference: bool,
    is_mut_ref: bool,
) -> ArgType {
    let rawtype;
    match typ {
        Type::Tuple(tup) if tup.elems.is_empty() => {
            rawtype = typ.to_token_stream();
            //argtype = ArgType::PlainData(typ.to_token_stream());
        }
        Type::Reference(typref) => {
            if typref.lifetime.is_some() {
                panic!(
                    "Method {}, argument {}: Specifying lifetimes is not supported.",
                    method_name, arg_name
                );
            }
            if is_reference {
                panic!("Method {}, argument {}: Method arguments cannot be reference to reference in Savefile-abi. Try removing a '&' from the type: {}", method_name, arg_name, typ.to_token_stream());
            }
            return parse_type(
                version,
                arg_name,
                &*typref.elem,
                method_name,
                &mut *name_generator,
                extra_definitions,
                true,
                typref.mutability.is_some(),
            );
        }
        Type::Tuple(tuple) => {
            if tuple.elems.len() > 3 {
                panic!("Savefile presently only supports tuples up to 3 members. Either change to using a struct, or file an issue on savefile!");
            }
            rawtype = tuple.to_token_stream();
        }
        Type::Slice(slice) => {
            if !is_reference {
                panic!(
                    "Method {}, argument {}: Slices must always be behind references. Try adding a '&' to the type: {}",
                    method_name,
                    arg_name,
                    typ.to_token_stream()
                );
            }
            if is_mut_ref {
                panic!("Method {}, argument {}: Mutable refernces are not supported by Savefile-abi, except for FnMut-trait objects. {}", method_name, arg_name, typ.to_token_stream());
            }
            return ArgType::SliceReference(slice.elem.to_token_stream());
        }
        Type::TraitObject(trait_obj) => {
            if !is_reference {
                panic!("Method {}, argument {}: Trait objects must always be behind references. Try adding a '&' to the type: {}", method_name, arg_name, typ.to_token_stream());
            }
            if trait_obj.dyn_token.is_some() {
                let type_bounds: Vec<_> = trait_obj
                    .bounds
                    .iter()
                    .filter_map(|x| match x {
                        TypeParamBound::Trait(t) => Some(
                            t.path
                                .segments
                                .iter()
                                .last()
                                .expect("Missing bounds of Box trait object"),
                        ),
                        TypeParamBound::Lifetime(_) => {
                            panic!(
                                "Method {}, argument {}: Specifying lifetimes is not supported.",
                                method_name, arg_name
                            );
                        }
                    })
                    .collect();
                if type_bounds.len() == 0 {
                    panic!("Method {}, argument {}, unsupported trait object reference. Only &dyn Trait is supported. Encountered zero traits.", method_name, arg_name);
                }
                if type_bounds.len() > 1 {
                    panic!("Method {}, argument {}, unsupported Box-type. Only &dyn Trait> is supported. Encountered multiple traits: {:?}", method_name, arg_name, trait_obj);
                }
                let bound = type_bounds.into_iter().next().expect("Internal error, missing bounds");

                if bound.ident == "Fn" || bound.ident == "FnMut" || bound.ident == "FnOnce" {
                    if bound.ident == "FnOnce" {
                        panic!(
                            "Method {}, argument {}, FnOnce is not supported. Maybe you can use FnMut instead?",
                            method_name, arg_name
                        );
                    }

                    if bound.ident == "FnMut" && !is_mut_ref {
                        panic!("Method {}, argument {}: When using FnMut, it must be referenced using &mut, not &. Otherwise, it is impossible to call.", method_name, arg_name);
                    }
                    let fn_decl = bound.to_token_stream();
                    match &bound.arguments {
                        PathArguments::Parenthesized(pararg) => {
                            //pararg.inputs
                            let temp_name =
                                Ident::new(&format!("{}_{}", &name_generator(), arg_name), Span::call_site());
                            emit_closure_helpers(
                                version,
                                temp_name.clone(),
                                pararg,
                                is_mut_ref,
                                extra_definitions,
                                bound.ident.clone(),
                            );
                            return ArgType::Fn(
                                temp_name,
                                fn_decl,
                                pararg.inputs.iter().map(|x| x.clone()).collect(),
                                is_mut_ref,
                            );
                        }
                        _ => {
                            panic!("Fn/FnMut arguments must be enclosed in parenthesis")
                        }
                    }
                } else {
                    return ArgType::TraitReference(bound.ident.clone(), is_mut_ref);
                }
            } else {
                panic!(
                    "Method {}, argument {}, reference to trait objects without 'dyn' are not supported.",
                    method_name, arg_name
                );
            }
        }
        Type::Path(path) => {
            let first_seg = path.path.segments.iter().next().expect("Missing path segments");
            if first_seg.ident == "Box" {
                match &first_seg.arguments {
                    PathArguments::AngleBracketed(ang) => {
                        let first_gen_arg = ang.args.iter().next().expect("Missing generic args of Box");
                        if ang.args.len() != 1 {
                            panic!("Method {}, argument {}. Savefile requires Box arguments to have exactly one generic argument, a requirement not satisfied by type: {:?}", method_name, arg_name, typ);
                        }
                        match first_gen_arg {
                            GenericArgument::Type(angargs) => match angargs {
                                Type::TraitObject(trait_obj) => {
                                    if is_reference {
                                        panic!("Method {}, argument {}: Reference to boxed trait object is not supported by savefile. Try using a regular reference to the box content instead.", method_name, arg_name);
                                    }
                                    let type_bounds: Vec<_> = trait_obj
                                        .bounds
                                        .iter()
                                        .filter_map(|x| match x {
                                            TypeParamBound::Trait(t) => Some(
                                                t.path
                                                    .segments
                                                    .iter()
                                                    .last()
                                                    .cloned()
                                                    .expect("Missing bounds of Box trait object")
                                                    .ident
                                                    .clone(),
                                            ),
                                            TypeParamBound::Lifetime(_) => None,
                                        })
                                        .collect();
                                    if type_bounds.len() == 0 {
                                        panic!("Method {}, argument {}, unsupported Box-type. Only Box<dyn Trait> is supported. Encountered zero traits in Box.", method_name, arg_name);
                                    }
                                    if type_bounds.len() > 1 {
                                        panic!("Method {}, argument {}, unsupported Box-type. Only Box<dyn Trait> is supported. Encountered multiple traits in Box: {:?}", method_name, arg_name, trait_obj);
                                    }
                                    if trait_obj.dyn_token.is_none() {
                                        panic!("Method {}, argument {}, unsupported Box-type. Only Box<dyn Trait> is supported.", method_name, arg_name)
                                    }
                                    let bound = type_bounds.into_iter().next().expect("Internal error, missing bounds");
                                    return ArgType::BoxedTrait(bound);
                                }
                                _ => {
                                    match parse_type(
                                        version,
                                        arg_name,
                                        angargs,
                                        method_name,
                                        &mut *name_generator,
                                        extra_definitions,
                                        is_reference,
                                        is_mut_ref,
                                    ) {
                                        ArgType::PlainData(_plain) => {
                                            rawtype = path.to_token_stream();
                                        }
                                        _ => {
                                            panic!(
                                                "Method {}, argument {}, unsupported Box-type: {:?}",
                                                method_name, arg_name, typ
                                            );
                                        }
                                    }
                                }
                            },
                            _ => {
                                panic!(
                                    "Method {}, argument {}, unsupported Box-type: {:?}",
                                    method_name, arg_name, typ
                                );
                            }
                        }
                    }
                    _ => {
                        panic!(
                            "Method {}, argument {}, unsupported Box-type: {:?}",
                            method_name, arg_name, typ
                        );
                    }
                }
            } else {
                rawtype = path.to_token_stream();
            }
        }
        _ => {
            panic!(
                "Method {}, argument {}, unsupported type: {:?}",
                method_name, arg_name, typ
            );
        }
    }
    if !is_reference {
        ArgType::PlainData(rawtype)
    } else {
        if is_mut_ref {
            panic!("Method {}, argument {}: Mutable references are not supported by Savefile-abi (except for FnMut-trait objects): {}", method_name, arg_name, typ.to_token_stream());
        }
        ArgType::Reference(rawtype)
    }
}

pub(super) fn generate_method_definitions(
    version: u32,
    trait_name: Ident,
    method_number: u16,
    method_name: Ident,
    ret_declaration: TokenStream, //May be empty, for ()-returns
    ret_type: TokenStream,
    receiver_is_mut: bool,
    args: Vec<(Ident, &Type)>,
    name_generator: &mut impl FnMut() -> String,
    extra_definitions: &mut Vec<TokenStream>,
) -> MethodDefinitionComponents {
    let method_name_str = method_name.to_string();

    let mut callee_trampoline_real_method_invocation_arguments: Vec<TokenStream> = vec![];
    let mut callee_trampoline_variable_declaration = vec![];
    let mut callee_trampoline_temp_variable_declaration = vec![];
    let mut callee_trampoline_variable_deserializer = vec![];
    let mut caller_arg_serializers = vec![];
    let mut caller_fn_arg_list = vec![];
    let mut metadata_arguments = vec![];

    for (arg_index, (arg_name, typ)) in args.iter().enumerate() {
        let argtype = parse_type(
            version,
            arg_name,
            *typ,
            &method_name,
            &mut *name_generator,
            extra_definitions,
            false,
            false,
        );

        //let num_mask = 1u64 << (method_number as u64);
        let temp_arg_name = Ident::new(&format!("temp_{}", arg_name), Span::call_site());
        let temp_arg_name2 = Ident::new(&format!("temp2_{}", arg_name), Span::call_site());
        match &argtype {
            ArgType::PlainData(_) => {
                callee_trampoline_real_method_invocation_arguments.push(quote! {#arg_name});
            }
            ArgType::Reference(_) => {
                callee_trampoline_real_method_invocation_arguments.push(quote! {&#arg_name});
                callee_trampoline_temp_variable_declaration.push(quote! {let #temp_arg_name;});
            }
            ArgType::SliceReference(_) => {
                callee_trampoline_real_method_invocation_arguments.push(quote! {&#arg_name});
                callee_trampoline_temp_variable_declaration.push(quote! {let #temp_arg_name;});
            }
            ArgType::BoxedTrait(_) => {
                callee_trampoline_real_method_invocation_arguments.push(quote! {#arg_name});
                callee_trampoline_temp_variable_declaration.push(quote! {let #temp_arg_name;});
            }
            ArgType::TraitReference(_, ismut) => {
                callee_trampoline_real_method_invocation_arguments.push(quote! {#arg_name});
                let mutsymbol = if *ismut {
                    quote!(mut)
                } else {
                    quote! {}
                };
                callee_trampoline_temp_variable_declaration.push(quote! {let #mutsymbol #temp_arg_name;});
            }
            ArgType::Fn(_, _, _, ismut) => {
                callee_trampoline_real_method_invocation_arguments.push(quote! {#arg_name});
                let mutsymbol = if *ismut {
                    quote!(mut)
                } else {
                    quote! {}
                };
                callee_trampoline_temp_variable_declaration.push(quote! {let #mutsymbol #temp_arg_name;});
                callee_trampoline_temp_variable_declaration.push(quote! {let #mutsymbol #temp_arg_name2;});
            }
        }
        callee_trampoline_variable_declaration.push(quote! {let #arg_name;});
        match &argtype {
            ArgType::Reference(arg_type) => {
                callee_trampoline_variable_deserializer.push(quote! {
                    if compatibility_mask&(1<<#arg_index) != 0 {
                        #arg_name = unsafe { &*(deserializer.read_raw_ptr::<#arg_type>()?) };
                    } else {
                        #temp_arg_name = <#arg_type as Deserialize>::deserialize(&mut deserializer)?;
                        #arg_name = &#temp_arg_name;
                    }
                });
                caller_arg_serializers.push(quote! {
                            if compatibility_mask&(1<<#arg_index) != 0 {
                                unsafe { serializer.write_raw_ptr(#arg_name as *const #arg_type).expect("Writing argument ref") };
                            } else {
                                #arg_name.serialize(&mut serializer).expect("Writing argument serialized");
                            }
                        });
            }
            ArgType::SliceReference(arg_type) => {
                callee_trampoline_variable_deserializer.push(quote! {
                    if compatibility_mask&(1<<#arg_index) != 0 {
                        #arg_name = unsafe { &*(deserializer.read_raw_ptr::<[#arg_type]>()?) };
                    } else {
                        #temp_arg_name = deserialize_slice_as_vec::<_,#arg_type>(&mut deserializer)?;
                        #arg_name = &#temp_arg_name;
                    }
                });
                caller_arg_serializers.push(quote! {
                            if compatibility_mask&(1<<#arg_index) != 0 {
                                unsafe { serializer.write_raw_ptr(#arg_name as *const [#arg_type]).expect("Writing argument ref") };
                            } else {
                                #arg_name.serialize(&mut serializer).expect("Writing argument serialized");
                            }
                        });
            }
            ArgType::PlainData(arg_type) => {
                callee_trampoline_variable_deserializer.push(quote! {
                    #arg_name = <#arg_type as Deserialize>::deserialize(&mut deserializer)?;
                });
                caller_arg_serializers.push(quote! {
                    #arg_name.serialize(&mut serializer).expect("Serializing arg");
                });
            }
            ArgType::BoxedTrait(trait_type) => {
                callee_trampoline_variable_deserializer.push(quote! {
                    // SAFETY
                    // Todo: Well, why exactly?
                    if compatibility_mask&(1<<#arg_index) == 0 {
                        panic!("Function arg is not layout-compatible!")
                    }
                    #temp_arg_name = unsafe { PackagedTraitObject::deserialize(&mut deserializer)? };
                    #arg_name = Box::new(unsafe { AbiConnection::from_raw_packaged(#temp_arg_name, Owning::Owned)? } );
                });
                caller_arg_serializers.push(quote! {
                            if compatibility_mask&(1<<#arg_index) == 0 {
                                panic!("Function arg is not layout-compatible!")
                            }
                            PackagedTraitObject::new::<dyn #trait_type>(#arg_name).serialize(&mut serializer).expect("PackagedTraitObject");
                        });
            }
            ArgType::TraitReference(trait_type, ismut) => {
                let mutsymbol = if *ismut {
                    quote! {mut}
                } else {
                    quote! {}
                };
                let newsymbol = if *ismut {
                    quote! {new_from_ptr}
                } else {
                    quote! {new_from_ptr}
                };
                callee_trampoline_variable_deserializer.push(quote! {
                            if compatibility_mask&(1<<#arg_index) == 0 {
                                panic!("Function arg is not layout-compatible!")
                            }
                            #temp_arg_name = unsafe { AbiConnection::from_raw_packaged(PackagedTraitObject::deserialize(&mut deserializer)?, Owning::NotOwned)? };
                            #arg_name = & #mutsymbol #temp_arg_name;
                        });
                caller_arg_serializers.push(quote! {
                            if compatibility_mask&(1<<#arg_index) == 0 {
                                panic!("Function arg is not layout-compatible!")
                            }
                            PackagedTraitObject::#newsymbol::<dyn #trait_type>( unsafe { std::mem::transmute(#arg_name) } ).serialize(&mut serializer).expect("PackagedTraitObject");
                        });
            }
            ArgType::Fn(temp_trait_type, _, args, ismut) => {
                let mutsymbol = if *ismut {
                    quote! {mut}
                } else {
                    quote! {}
                };
                let mutorconst = if *ismut {
                    quote! {mut}
                } else {
                    quote! {const}
                };
                let newsymbol = if *ismut {
                    quote! {new_from_ptr}
                } else {
                    quote! {new_from_ptr}
                };

                let temp_trait_name_wrapper = Ident::new(&format!("{}_wrapper", temp_trait_type), Span::call_site());

                let typedarglist: Vec<TokenStream> = args
                    .iter()
                    .enumerate()
                    .map(|(idx, typ)| {
                        let id = Ident::new(&format!("x{}", idx), Span::call_site());
                        quote! {#id : #typ}
                    })
                    .collect();

                let arglist: Vec<Ident> = (0..args.len())
                    .map(|idx| {
                        let id = Ident::new(&format!("x{}", idx), Span::call_site());
                        id
                    })
                    .collect();
                callee_trampoline_variable_deserializer.push(quote! {
                    if compatibility_mask&(1<<#arg_index) == 0 {
                        panic!("Function arg is not layout-compatible!")
                    }

                    #temp_arg_name = unsafe { AbiConnection::<#temp_trait_type>::from_raw_packaged(PackagedTraitObject::deserialize(&mut deserializer)?, Owning::NotOwned)? };
                    #temp_arg_name2 = |#(#typedarglist,)*| {#temp_arg_name.docall(#(#arglist,)*)};
                    #arg_name = & #mutsymbol #temp_arg_name2;
                });
                caller_arg_serializers.push(quote! {
                    if compatibility_mask&(1<<#arg_index) == 0 {
                        panic!("Function arg is not layout-compatible!")
                    }

                    let #mutsymbol temp = #temp_trait_name_wrapper { func: #arg_name as *#mutorconst _ };
                    let #mutsymbol temp : *#mutorconst (dyn #temp_trait_type+'_) = &#mutsymbol temp as *#mutorconst _;
                    PackagedTraitObject::#newsymbol::<(dyn #temp_trait_type+'_)>( unsafe { std::mem::transmute(temp)} ).serialize(&mut serializer).expect("PackagedTraitObject");
                });
            }
        }
        match &argtype {
            ArgType::Reference(arg_type) => {
                caller_fn_arg_list.push(quote! {#arg_name : &#arg_type});
                metadata_arguments.push(quote! {
                    AbiMethodArgument {
                        schema: <#arg_type as WithSchema>::schema(version),
                        can_be_sent_as_ref: true
                    }
                })
            }
            ArgType::SliceReference(arg_type) => {
                caller_fn_arg_list.push(quote! {#arg_name : &[#arg_type]});
                metadata_arguments.push(quote! {
                    AbiMethodArgument {
                        schema: <[#arg_type] as WithSchema>::schema(version),
                        can_be_sent_as_ref: true
                    }
                })
            }
            ArgType::PlainData(arg_type) => {
                caller_fn_arg_list.push(quote! {#arg_name : #arg_type});
                metadata_arguments.push(quote! {
                    AbiMethodArgument {
                        schema: <#arg_type as WithSchema>::schema(version),
                        can_be_sent_as_ref: false
                    }
                })
            }
            ArgType::BoxedTrait(trait_name) => {
                caller_fn_arg_list.push(quote! {#arg_name : Box<dyn #trait_name>});
                metadata_arguments.push(quote! {
                    AbiMethodArgument {
                        schema: Schema::BoxedTrait(<dyn #trait_name as AbiExportable>::get_definition(version)),
                        can_be_sent_as_ref: true
                    }
                })
            }
            ArgType::TraitReference(trait_name, ismut) => {
                if *ismut {
                    caller_fn_arg_list.push(quote! {#arg_name : &mut dyn #trait_name });
                } else {
                    caller_fn_arg_list.push(quote! {#arg_name : &dyn #trait_name });
                }

                metadata_arguments.push(quote! {
                    AbiMethodArgument {
                        schema: Schema::BoxedTrait(<dyn #trait_name as AbiExportable>::get_definition(version)),
                        can_be_sent_as_ref: true,
                    }
                })
            }
            ArgType::Fn(temp_trait_name, fndef, _, ismut) => {
                if *ismut {
                    caller_fn_arg_list.push(quote! {#arg_name : &mut dyn #fndef });
                } else {
                    caller_fn_arg_list.push(quote! {#arg_name : &dyn #fndef });
                }
                //let temp_trait_name_str = temp_trait_name.to_string();
                metadata_arguments.push(quote! {
                            {
                                AbiMethodArgument {
                                    schema: Schema::FnClosure(#ismut, <dyn #temp_trait_name as AbiExportable >::get_definition(version)),
                                    can_be_sent_as_ref: true,
                                }
                            }
                        })
            }
        }
    }

    let callee_real_method_invocation_except_args;
    if receiver_is_mut {
        callee_real_method_invocation_except_args =
            quote! { unsafe { &mut *trait_object.as_mut_ptr::<dyn #trait_name>() }.#method_name };
    } else {
        callee_real_method_invocation_except_args =
            quote! { unsafe { &*trait_object.as_const_ptr::<dyn #trait_name>() }.#method_name };
    }

    //let receiver_mut_str = receiver_mut.to_string();
    let receiver_mut = if receiver_is_mut {
        quote!(mut)
    } else {
        quote! {}
    };
    let caller_method_trampoline = quote! {
        fn #method_name(& #receiver_mut self, #(#caller_fn_arg_list,)*) #ret_declaration {
            let info: &AbiConnectionMethod = &self.template.methods[#method_number as usize];

            let Some(callee_method_number) = info.callee_method_number else {
                panic!("Method '{}' does not exist in implementation.", info.method_name);
            };

            let mut result_buffer: MaybeUninit<Result<#ret_type,SavefileError>> = MaybeUninit::<Result<#ret_type,SavefileError>>::uninit();
            let compatibility_mask = info.compatibility_mask;

            let mut data = FlexBuffer::new();
            let mut serializer = Serializer {
                writer: &mut data,
                file_version: self.template.effective_version,
            };
            serializer.write_u32(self.template.effective_version).unwrap();
            #(#caller_arg_serializers)*

            (self.template.entry)(AbiProtocol::RegularCall {
                trait_object: self.trait_object,
                compatibility_mask: compatibility_mask,
                method_number: callee_method_number,
                effective_version: self.template.effective_version,
                data: data.as_ptr() as *const u8,
                data_length: data.len(),
                abi_result: &mut result_buffer as *mut MaybeUninit<Result<#ret_type,SavefileError>> as *mut (),
                receiver: abi_result_receiver::<#ret_type>,
            });
            let resval = unsafe { result_buffer.assume_init() };

            resval.expect("Unexpected panic in invocation target")
        }
    };

    let method_metadata = quote! {
        AbiMethod {
            name: #method_name_str.to_string(),
            info: AbiMethodInfo {
                return_value: <#ret_type as WithSchema>::schema(version),
                arguments: vec![ #(#metadata_arguments,)* ],
            }
        }
    };

    let callee_method_trampoline = quote! {
        #method_number => {
            #(#callee_trampoline_variable_declaration)*
            #(#callee_trampoline_temp_variable_declaration)*

            #(#callee_trampoline_variable_deserializer)*

            let ret = #callee_real_method_invocation_except_args( #(#callee_trampoline_real_method_invocation_arguments,)* );

            let mut slow_temp = FlexBuffer::new();
            let mut serializer = Serializer {
                writer: &mut slow_temp,
                file_version: #version,
            };
            serializer.write_u32(effective_version)?;
            match ret.serialize(&mut serializer)
            {
                Ok(()) => {
                    let outcome = RawAbiCallResult::Success {data: slow_temp.as_ptr(), len: slow_temp.len()};
                    receiver(&outcome as *const _, abi_result);
                }
                Err(err) => {
                    let err_str = format!("{:?}", err);
                    let outcome = RawAbiCallResult::AbiError(AbiErrorMsg{error_msg_utf8: err_str.as_ptr(), len: err_str.len()});
                    receiver(&outcome as *const _, abi_result)
                }
            }

        }

    };
    MethodDefinitionComponents {
        method_metadata,
        callee_method_trampoline,
        caller_method_trampoline,
    }
}
