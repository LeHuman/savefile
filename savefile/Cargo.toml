[package]
name = "savefile"
version = "0.14.3"
authors = ["Anders Musikka <anders@andersmusikka.se>"]
documentation = "https://docs.rs/savefile/"
homepage = "https://github.com/avl/savefile/"

exclude = [
    "*.bin"
]

description = "Simple, convenient, fast, versioned, binary serialization/deserialization library. Works on stable rust, but provides much better performance with nightly."

readme = "../README.md"

keywords = ["serialization", "deserialization", "introspection"]

categories = ["encoding"]

license = "MIT/Apache-2.0"

edition = "2018"

[features]
default = ["indexmap", "arrayvec", "smallvec", "bit-vec", "parking_lot","bit-set"]
bit-set = ["dep:bit-set","bit-vec"]

# Enable this to reduce risk of crashing on corrupt input. Provides sanity checks for sizes of objects.
# This is mostly to be able to run fuzzers against the deserializers without them being guaranteed to easily find out-of-memory crashes.
size_sanity_checks = []
# Use features only available on the nightly rust-compiler.
# Enabling this means serialization of vectors of simple types will be much faster.
# It also enables serialization of arbitrary size arrays (arbitrary size Vec always work, regardless of features)
nightly=[]

compression = ["bzip2"]

encryption = ["ring", "rand"]

[dependencies]
bit-vec = { version = "0.6", optional = true}
arrayvec = { version = "0.7", optional = true}
smallvec = { version = "1.0", optional = true}
indexmap = { version = "1.6", optional = true}
parking_lot = { version = "0.12", optional = true }
ring = {version = "0.16.9", optional = true}
rand = { version = "0.8", optional = true}
bzip2 = {version = "0.4.4", optional = true}
bit-set = {version = "0.5", optional = true}
rustc-hash = {version = "1.1", optional = true}

byteorder = "1.2"


[dev-dependencies]
savefile-derive = { path="../savefile-derive" }

[build-dependencies]
rustc_version="0.2"



