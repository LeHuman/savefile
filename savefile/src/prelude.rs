pub use {
    super::introspect_item, super::load, super::load_file, super::load_file_noschema, super::load_from_mem, super::load_noschema, super::save,
    super::save_file, super::save_file_noschema, super::save_noschema, super::save_to_mem, super::Canary1, super::Deserialize, super::Deserializer, super::Field, super::Introspect, super::IntrospectItem, super::IntrospectedElementKey,
    super::IntrospectionResult, super::Introspector, super::IntrospectorNavCommand, super::Removed, super::ReprC, super::SavefileError, super::Schema, super::SchemaEnum,
    super::SchemaPrimitive, super::SchemaStruct, super::Serialize, super::Serializer, super::Variant, super::WithSchema, super::IsReprC
};

#[cfg(feature="ring")]
pub use super::{CryptoReader, CryptoWriter, save_encrypted_file, load_encrypted_file};
