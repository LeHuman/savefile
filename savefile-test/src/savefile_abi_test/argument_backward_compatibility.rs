use savefile::{AbiRemoved, SavefileError};
use savefile_abi::{AbiConnection, AbiExportable, verify_compatiblity};
use savefile_abi::RawAbiCallResult::AbiError;
use savefile_abi_test::argument_backward_compatibility::v1::{ArgInterfaceV1, Implementation1};
use savefile_abi_test::argument_backward_compatibility::v2::{ArgInterfaceV2, Implementation2};
use savefile_derive::Savefile;


mod v1 {
    #[derive(Savefile)]
    pub struct Argument {
        pub data1: u32,
        pub data2: u32,
    }
    #[savefile_abi_exportable(version=0)]
    pub trait ArgInterfaceV1 {
        fn sums(&self, a: Argument, b: Argument) -> u32;
    }
    #[derive(Default)]
    pub struct Implementation1 {}


    impl ArgInterfaceV1 for Implementation1 {
        fn sums(&self, a: Argument, b: Argument) -> u32 {
            a.data1+a.data2+b.data1+b.data2
        }
    }
}

mod v2 {
    use savefile::AbiRemoved;
    use savefile_derive::{Savefile};
    use ::savefile::prelude::*;

    #[derive(Savefile)]
    pub struct ArgArgument {
        #[savefile_versions="0..0"]
        pub data1: AbiRemoved<u32,>,
        pub data2: u32,
        #[savefile_versions="1.."]
        pub data3: u32,
    }
    #[savefile_abi_exportable(version=1)]
    pub trait ArgInterfaceV2 {
        fn sums(&self, a: ArgArgument, b: ArgArgument) -> u32;
    }

    #[derive(Default)]
    pub struct Implementation2 {}
    impl ArgInterfaceV2 for Implementation2 {
        fn sums(&self, a: ArgArgument, b: ArgArgument) -> u32 {
            a.data3+a.data2+b.data2+b.data3
        }
    }
}



#[test]
pub fn test_abi_schemas_get_def() {
    let exportable = <dyn ArgInterfaceV2 as AbiExportable>::get_definition(0);
    insta::assert_yaml_snapshot!(exportable);
}

#[test]
#[cfg(not(miri))]
pub fn test_backward_compatibility() -> Result<(), SavefileError> {
    verify_compatiblity::<dyn ArgInterfaceV2>("schemas")
}

#[test]
pub fn test_caller_has_older_version() {
    let iface2 : Box<dyn ArgInterfaceV2> = Box::new(Implementation2{});
    let conn1 = unsafe { AbiConnection::<dyn ArgInterfaceV1>::from_boxed_trait_for_test(<dyn ArgInterfaceV2 as AbiExportable>::ABI_ENTRY, iface2 ) }.unwrap();

    assert_eq!(conn1.sums(
        v1::Argument {
            data1: 2,
            data2: 3,
        },
        v1::Argument {
            data1: 4,
            data2: 5,
        },
    ), 8); //Because implementation expects data2 and data3, but we're only sending data2.
}

#[test]
pub fn test_caller_has_newer_version() {
    let iface1 : Box<dyn ArgInterfaceV1> = Box::new(Implementation1{});
    let conn1 = unsafe { AbiConnection::<dyn ArgInterfaceV2>::from_boxed_trait_for_test(<dyn ArgInterfaceV1 as AbiExportable>::ABI_ENTRY, iface1 ) }.unwrap();

    assert_eq!(conn1.sums(
        v2::ArgArgument {
            data1: AbiRemoved::new(),
            data2: 1,
            data3: 2
        },
        v2::ArgArgument {
            data1: AbiRemoved::new(),
            data2: 3,
            data3: 4
        },
    ), 4); //Because implementation expects data1 and data2, but we're only sending data2.
}
