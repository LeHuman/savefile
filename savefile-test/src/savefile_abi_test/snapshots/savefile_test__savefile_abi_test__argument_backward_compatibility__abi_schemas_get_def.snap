---
source: savefile-test/src/savefile_abi_test/argument_backward_compatibility.rs
expression: exportable
---
name: ArgInterfaceV2
methods:
  - name: sums
    info:
      return_value:
        Primitive: schema_u32
      arguments:
        - schema:
            Struct:
              dbg_name: ArgArgument
              size: 8
              alignment: 4
              fields:
                - name: data1
                  value:
                    Primitive: schema_u32
                  offset: ~
                - name: data2
                  value:
                    Primitive: schema_u32
                  offset: 0
          can_be_sent_as_ref: false
        - schema:
            Struct:
              dbg_name: ArgArgument
              size: 8
              alignment: 4
              fields:
                - name: data1
                  value:
                    Primitive: schema_u32
                  offset: ~
                - name: data2
                  value:
                    Primitive: schema_u32
                  offset: 0
          can_be_sent_as_ref: false
  - name: enum_arg
    info:
      return_value:
        Primitive:
          schema_string: CapacityDataLength
      arguments:
        - schema:
            Enum:
              dbg_name: EnumArgument
              variants:
                - name: Variant1
                  discriminant: 0
                  fields: []
                - name: Variant2
                  discriminant: 1
                  fields: []
              discriminant_size: 1
              has_explicit_repr: false
              size: 1
              alignment: 1
          can_be_sent_as_ref: false
